### DISCUSSION TITLE
Unable to perform parameter optimization for sugarcane

### DISCUSSION THREAD

**Discussion by Clzuy** [APSIMInitiative/ApsimX#9372] [Q&A]
Hi,

Based on the R package apsimx, I followed the parameter optimization example provided at https://femiguez.github.io/apsimx-docs/optim-apsim.html#optim_apsim to optimize parameters for the sugarcane crop, but I encountered some issues during the optimization process.

Initially, when using the default sugarcane type q138 provided by the system, I encountered the following error:

Error in inspect_apsimx_replacement("sugarRuan1004Opt.apsimx", src.dir = extd.dir, : More than one subchild found. Make it unique (see regular expressions)

Therefore, I created a new cultivar named clzuy and included it. This resulted in the following message:

> inspect_apsimx_replacement("sugarRuan1004Opt.apsimx", src.dir = extd.dir, + node = "Sugarcane", + node.child = "Cultivars", + node.subchild = "clzuy", + verbose = FALSE) These positions matched Models.Core.Folder 1 3 Error: Multiple root nodes found. Please provide a position I specified a position root = list("Models.Core.Folder", 1) and then it ran successfully: $type : Models.PMF.Cultivar, Models : [Sugarcane].cult.leaf_size = 1500,55000,55000 : [Sugarcane].cult.leaf_size_no = 1,14,20 : [Sugarcane].cult.cane_fraction =0.7 : [Sugarcane].cult.sucrose_fraction_stalk = 1,0.61 : [Sugarcane].cult.stress_factor_stalk = 0.2,1 : [Sugarcane].cult.sucrose_delay = 600 : [Sugarcane].cult.min_sstem_sucrose = 1500 : [Sugarcane].cult.min_sstem_sucrose_redn = 10 : [Sugarcane].cult.tt_emerg_to_begcane = 1900 : [Sugarcane].cult.tt_begcane_to_flowering = 6000 : [Sugarcane].cult.tt_flowering_to_crop_end = 2000 : [Sugarcane].cult.green_leaf_no = 13 : [Sugarcane].cult.tillerf_leaf_size = 1,1,1.5,1,1 : [Sugarcane].cult.tillerf_leaf_size_no = 1,4,10,16,26 Command Name : clzuy Enabled : TRUE ReadOnly : FALSE

I then wrote the parameter paths as follows, based on the example:

pp1 <- "Sugarcane.Cultivars.clzuy.cane_fraction" pp2 <- "Sugarcane.Cultivars.clzuy.sucrose_fraction_stalk" pp3 <- "Sugarcane.Cultivars.clzuy.min_sstem_sucrose" pp4 <- "Sugarcane.Cultivars.clzuy.tt_emerg_to_begcane" pp5 <- "Sugarcane.Cultivars.clzuy.green_leaf_no" However, when calling optim_apsimx, I encountered the following error:
I am unable to resolve this error. From my comparison, I found that the sugarcane module is directly migrated from APSIM 7.10 to APSIMX, and its parameters differ from other crops in that they do not have the .FixedValue suffix. If I try to add .FixedValue to the parameters, it results in an immediate error.
APSIMX VERSION-2024.9.7589.0 (genotype parameters for Sugarcane update on version-2024.07.7562)
I have compressed and attached the relevant files. Could you please suggest any solutions?

Frustratingly, I continued to replaced the parameter optimization R package with CroptimizR and followed the steps from [this link](https://sticsrpacks.github.io/CroptimizR/articles/ApsimX_parameter_estimation_simple_case.html), but still couldn't find the path to modify the sugarcane cultivar parameters.

```r
param_info <-
  list(lb = c(.Simulations.Replacements.Sugarcane.Cultivars.q138.cane_fraction = 0.65,
              .Simulations.Replacements.Sugarcane.Cultivars.q138.min_sstem_sucrose = 400),
       ub = c(.Simulations.Replacements.Sugarcane.Cultivars.q138.cane_fraction = 0.80,
              .Simulations.Replacements.Sugarcane.Cultivars.q138.min_sstem_sucrose = 1500
              ))
```

"System.Exception: Invalid path: .Simulations.Replacements.Sugarcane.Cultivars.q138.cane_fraction"    

This is so disheartening; I can't proceed with my experiment. Can anyone provide assistance?
Thank you!

[exam.zip](https://github.com/user-attachments/files/17265020/exam.zip)

Best regards,
Zongheng Li

**Comment by i-heis** [comment]
Hi did you resolve the issue yet? I had a quick look at your apsim file and it seems like you are trying to optimize plant parameters that you have defined in the cultivar sheet. That does not work with the croptimizr. If you wish to optimize certain but not all plant parameters, you need to include all relevant parameters (if they differ from the default cultivar parameters in apsim) in the cultivar sheet and exclude the ones that you want to optimize. 
I hope this helps.

**Reply by Clzuy** [reply]
Thank you so much for your response! I’ve managed to solve the issue and successfully graduated—haha! Really appreciate your help throughout. All the best！
